enhancements/ingress/wildcard-routing.md

```diff
@@ -0,0 +1,189 @@
+  - "@openshift/api-approvers"
+creation-date: 2020-03-19
+last-updated: 2020-03-19
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Wildcard Routing Policy
+
+Application developers would like to expose a service outside the cluster by multiple names without specify each one
+individually.
+
+In OpenShift version 4, wildcard routing is not supported. The `Route` resource supports setting a wildcard policy but
+no API exists to enable this feature on an ingress controller (i.e. router).
+
+OpenShift version 3 supports this feature through the `ROUTER_ALLOW_WILDCARD_ROUTES` environment variable when
+configuring the OpenShift Router directly. Users of this feature are blocked from upgrading to OpenShift 4 unless they
+discontinue use of the feature.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+See the [section][#wildcard-routing-policy]
+
+## Motivation
+
+See the [section][#wildcard-routing-policy]
+
+### Goals
+
+1. Define an API field that allows a cluster administrator to enable wildcard routing for an ingress controller.
+
+### Non-Goals
+
+1. Redesign how wildcard routing is implemented in [OpenShift Router](https://github.com/openshift/router).
+2. Introduce a separate API for managing wildcard routing.
+
+## Proposal
+
+Introduce a `WildcardRouting` field to the [IngressController](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go)
+resource that allows administrators to enable or disable wildcard routing.
+
+### User Stories
+
+### Implementation Details/Notes/Constraints [optional]
+
+The [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) will use `WildcardRouting` to configure
+the environment variable `ROUTER_ALLOW_WILDCARD_ROUTES` in the [OpenShift Router](https://github.com/openshift/router).
+
+The default behavior should be to disallow wildcard routing, which is backwards compatible with existing
+`IngressController` resources.
+
+### Risks and Mitigations
+
+Addition of the new field requires no migration, as the default is not changing.
+
+## Design Details
+
+### New API Field
+
+`WildcardRouting` enum definition.
+
+```go
+// (Existing fields omitted, only newly proposed fields are visible)
+type RouteAdmissionPolicy struct {
+  // wildcardRouting enables or disables wildcard routing support for
+  // the ingress controller. WildcardRouting works in conjunction with
+  // the WildcardPolicy of routes [1] claimed by the ingress controller.
+  //
+  // [1] https://github.com/openshift/api/blob/master/route/v1/types.go
+  //
+  // WildcardRouting can not be updated.
+  //
+  // If empty, defaults to "Disabled".
+  //
+  WildcardRouting *WildcardRoutingType `json:"wildcardRouting,omitempty"`
```

Comment:
Current name and enum values feel like a boolean when it seems like a policy to me... another name to consider would be `WildcardPolicy`

Reply:
:-) so funny... I kept going back and forth with `WildcardPolicy`, `WildcardRoutePolicy` and `WildcardRoutingPolicy` as options for naming the field. I dropped the Policy portion of the name b/c I thought it was redundant with `RouteAdmissionPolicy`. @ironcladlou do you feel a policy within a policy makes sense?

Reply:
I don't feel strongly and think what you have here is fine. The word "routing" does seem to imply coupling to Route (does this apply to Ingress?) but I'm not sure that matters?

Reply:
@ironcladlou let's see if @Miciah and @frobware have any feedback on the name.

Reply:
> The word "routing" does seem to imply coupling to Route (does this apply to Ingress?) but I'm not sure that matters?

@ironcladlou unless I'm missing something, Ingress v1 does not support [wildcards](https://github.com/kubernetes/kubernetes/issues/41881).

Reply:
How about `spec.routeAdmission.wildcard: Allow|Deny`?

Reply:
I'm using `WildcardPolicy` in my latest commit.